src/utils/Cypher.ts

```diff
@@ -26,19 +32,79 @@ export default class Cypher {
     return this.configuration;
   }
 
-  public static encrypt(data: string): string {
+  public static async detectConfigurationKey(tenantID: string): Promise<void> {
+
+    const configCryptoKey: string = Configuration.getCryptoConfig().key;
+    const keySettings = await SettingStorage.getCryptoKeySettings(tenantID);
+    let cryptoKeySetting: CryptoKeySetting;
+
+    // Check if Crypto Key settings exists
+    if (keySettings) {
+      // Detect new config Cypher key
+      if (keySettings.cryptoKey.newKey && keySettings.cryptoKey.newKey !== configCryptoKey) {
+        cryptoKeySetting = {
+          oldKey: keySettings.cryptoKey.newKey
+        } as CryptoKeySetting;
+      } else if (!keySettings.cryptoKey.newKey && keySettings.cryptoKey.oldKey !== configCryptoKey) {
+        // If no newKey exist, detect oldKey change
+        cryptoKeySetting = {
+          oldKey: keySettings.cryptoKey.oldKey
+        } as CryptoKeySetting;
+      }
+      // If key change detected, prepare KeySettings change
+      if (cryptoKeySetting) {
+        await SettingStorage.deleteSetting(tenantID, keySettings.id);
+        cryptoKeySetting.newKey = configCryptoKey;
+
+        // Key migration of senitive data
+        this.cryptoKey = cryptoKeySetting;
+      }
+    } else {
+      // Create New Config Crypto Key in Tenant Settings
+      cryptoKeySetting = {
+        oldKey: configCryptoKey
+      } as CryptoKeySetting;
+    }
+
+    if (cryptoKeySetting) {
+      const keySettingToSave = {
+        identifier: TenantComponents.CRYPTO_KEY,
+        type: KeySettingsType.CRYPTO_KEY,
+        cryptoKey: cryptoKeySetting
+      } as KeySettings;
+      await SettingStorage.saveCryptoKeySettings(tenantID, keySettingToSave);
+    }
+  }
+
+  public static async migrateSensitiveData(tenantID: string, setting: SettingDB): Promise<void> {
+    if (this.cryptoKey) {
+      this.decryptSensitiveDataInJSON(setting, this.cryptoKey.oldKey);
+      this.encryptSensitiveDataInJSON(setting, this.cryptoKey.newKey);
+      await SettingStorage.saveSettings(tenantID, setting);
+    }
+  }
+
+  public static async migrateAllSensitiveData(tenantID: string, settings: SettingDB[]): Promise<void> {
+    for (const setting of settings) {
+      await this.migrateSensitiveData(tenantID, setting);
+    }
+  }
+
+  public static encrypt(data: string, key?: string): string {
```

Comment:
I think you should make the key argument required here. 
That way we know from reading the code calling that method which key is gonna be used.

Reply:
There are quite a few dependencies which had to be modified if we make key required instead of optional. I will commit changes and you can look to see if it's fine or if to keep key optional to not affect all of the other places.